common/featureset.cpp

```diff
@@ -0,0 +1,55 @@
+#include "featureset.h"
+#include <QMap>
+#include <QDebug>
+
+FeatureSet::FeatureSet()
+{
+
+}
+
+QMap<QString, bool> FeatureSet::getDefaultFeatureList() {
+    initalizeFeatureList(featureList);
+    return featureList;
+}
+
+void FeatureSet::initalizeFeatureList(QMap<QString, bool> &featureList){
+    featureList.insert("client_id", false);
+    featureList.insert("client_ver", false);
+    featureList.insert("feature_set", false);
+}
+
+void FeatureSet::enableRequiredFeature(QMap<QString, bool> &featureList, QString featureName){
+    if (featureList.contains(featureName))
+        featureList.insert(featureName,true);
+}
+
+void FeatureSet::disableRequiredFeature(QMap<QString, bool> &featureList, QString featureName){
+    if (featureList.contains(featureName))
+        featureList.insert(featureName,false);
+}
+
+QMap<QString, bool> FeatureSet::addFeature(QMap<QString, bool> &featureList, QString featureName, bool isFeatureRequired){
+    featureList.insert(featureName,isFeatureRequired);
+    return featureList;
+}
+
+QMap<QString, bool> FeatureSet::identifyMissingFeatures(QMap<QString, bool> suppliedFeatures, QMap<QString, bool> requiredFeatures){
+    QMap<QString, bool> missingList;
+    QMap<QString, bool>::iterator i;
+    for (i = requiredFeatures.begin(); i != requiredFeatures.end(); ++i) {
+        if (!suppliedFeatures.contains(i.key())) {
+            missingList.insert(i.key(), i.value());
+        }
+    }
+    return missingList;
+}
```

Comment:
Oh, I'm not sure. I think it should work, the docs say it's an STL-style
iterator. I haven't tested it.

On Tue, Sep 1, 2015 at 12:15 PM woogerboy21 <notifications@github.com>
wrote:

> In common/featureset.cpp
> <https://github.com/Cockatrice/Cockatrice/pull/1439#discussion_r38438340>:
>
> > +
> > +QMap<QString, bool> FeatureSet::addFeature(QMap<QString, bool> &featureList, QString featureName, bool isFeatureRequired){
> > +    featureList.insert(featureName,isFeatureRequired);
> > +    return featureList;
> > +}
> > +
> > +QMap<QString, bool> FeatureSet::identifyMissingFeatures(QMap<QString, bool> suppliedFeatures, QMap<QString, bool> requiredFeatures){
> > +    QMap<QString, bool> missingList;
> > +    QMap<QString, bool>::iterator i;
> > +    for (i = requiredFeatures.begin(); i != requiredFeatures.end(); ++i) {
> > +        if (!suppliedFeatures.contains(i.key())) {
> > +            missingList.insert(i.key(), i.value());
> > +        }
> > +    }
> > +    return missingList;
> > +}
>
> Right, but isn't your std:any_of expecting the begin / end values to be a
> std value type?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Cockatrice/Cockatrice/pull/1439/files#r38438340>.
>